projects/krasnor/proposal/README.md

```diff
@@ -0,0 +1,53 @@
+
+## Proposal
+
+![](assets/subdivision_example.jpg)
+
+​		*A cube (left) subdivided with the Catmull-Clark subdivision algorithm after 1 (middle) and 2 (right) iterations. Example rendered with [Blender](https://www.blender.org/).*
+
+I want to implement the Catmull-Clark subdivision algorithm [Catmull et al. 1978] for a halfedge-mesh structure [Kettner 2004]. During initial research I noticed that the algorithm can convert triangle faces into quadrilateral faces. Thus the halfedge structure must have quadrilateral face support. Additionally the visualization will support the user to compare the original mesh and subdivided mesh (along with additional information, such as vertex count) by displaying both versions at the same time next to each other. As remeshing is part of geometry processing this project is relevant to the lecture. 
+
+I will use HTML5, Javascript, [Node.js](https://nodejs.org/) and the library [three.js](http://threejs.org/) for this project. 
+
+The implementation of the halfedge structure will be based on the already known code from the exercises and lecture, but will be extended to be able to handle quadrilateral faces. The Catmull-Clark subdivision algorithm will be based on the original paper [Catmull et al. 1978], but be adjusted to use the halfedge structure instead of a matrix representation for the mesh. While the 3D-modelling software [Blender](https://www.blender.org/) also implements the Catmull-Clark subdivision algorithm with a halfedge structure it is written in C++, my implementation will be written in JavaScript and three.js. Thus the project is not based on an already existing code solution.
+
+## Implementation
+
+The project will be implemented by the usage of HTML5, JavaScript, [Node.js](https://nodejs.org/) and [three.js](http://threejs.org/). 
+
+Milestones will be:
+
+1. Implementing halfedge connectivity to support quadrilateral faces
+2. Adjust calculations for normals
+3. Get three.js to handle the quadrilateral Mesh (as three.js dropped Face4 support)
+4. Implement the Catmull–Clark Subdivision algorithm
+5. Add support to compare the original mesh and subdivided mesh, along with additional information (such as vertex count)
+
+Bigger challenges may encountered in Milestones 1, 3 and 4:
+
+- Milestone 1: a lot of code has to be adjusted to enable the support for quadrilateral faces, checking validity and debugging the generated mesh will be difficult.
+- Milestone 3: Three.js does not support rendering of quadrilateral faces, thus another way has to be found to still render the halfedge structure with three.js. 
+  As this is mostly a visual problem and OpenGl/WebGL typically would render the quadrilateral face as a pair of triangles my current approach will be to triangulate the mesh and overlay an wireframe material, which only renders the edges of the quadrilateral mesh version. This way the quadrilateral mesh structure can still be observed.
```

Comment:
I guess there are no existing hardware support rendering ngons directly(?). If you check the source code of blender, you will realize that their approach is the same. However, for the case where a face is not flat, the shading of the face will be carefully handled. For instance: Quad ABCD is equivalent to triangle ABC and triangle ACD, or ABD and BCD (depends on your implementation), but the connecting virtual edge does not cause it looks like triangles. A naive but working approach is to use an averaged face normal of that pair of triangles under Flat/Phong shaidng to compute the actual color.